### PR TITLE
Restore quantized RNN to master

### PR DESCRIPTION
**Pull request by bgawrych**
## Description ##
Restoring qunatized RNN: https://github.com/apache/incubator-mxnet/pull/20759

### CODE REVIEW

**Comment by bgawrych** [comment]
@mxnet-bot run ci [centos-cpu, unix-cpu, website, windows-gpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [unix-cpu, windows-gpu, clang]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [windows-gpu]

File: src/operator/nn/dnnl/dnnl_rnn.cc [hunk 1]
```diff
@@ -449,11 +482,19 @@ void DNNLRnnForward::SetNewDataMem(void* x,
   auto& cpu_engine      = CpuEngine::Get()->get_engine();
   dnnl_args_map_t& args = net_args_;
 
+  int src_dtype = dtype;
+  int dst_dtype = dtype;
+  if (param_.quantized) {
+    src_dtype = mshadow::kUint8;
```
Comment (bartekkuncer) [inline src/operator/nn/dnnl/dnnl_rnn.cc:488]: RNN does not support int8?

File: src/operator/quantization/dnnl/dnnl_quantized_rnn.cc [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+namespace op {
+
+std::vector<float> GetDNNLRnnWeightsQParams(const DNNLRnnFullParam& full_param, float* w_ptr) {
+  const int nthreads            = mxnet::engine::OpenMP::Get()->GetRecommendedOMPThreadCount();
+  const RNNParam& default_param = full_param.default_param;
+  const LayerParamVector& layer_params = full_param.layer_params;
+
+  const DNNLRnnLayerParam& layer_param0 = layer_params.at(0);
+  const size_t w_size0                  = layer_param0.single_w_size;
+  const size_t wx_size0                 = 4 * layer_param0.state_size * layer_param0.input_size;
```
Comment (bartekkuncer) [inline src/operator/quantization/dnnl/dnnl_quantized_rnn.cc:41]: Magic number.

File: src/operator/quantization/quantized_rnn-inl.h [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file quantized_rnn-inl.h
```
Comment (bartekkuncer) [inline src/operator/quantization/quantized_rnn-inl.h:21]: Is this file necessary?

File: src/operator/nn/dnnl/dnnl_rnn.cc (hunk 1 above)
Comment (bgawrych) [reply src/operator/nn/dnnl/dnnl_rnn.cc:488]: Now it does - seems like int8 is supported since 2.3 (https://oneapi-src.github.io/oneDNN/v2.2/dev_guide_rnn.html / https://oneapi-src.github.io/oneDNN/v2.3/dev_guide_rnn.html)

New task can be opened for int8 implementation and checking if it is worth

File: src/operator/quantization/quantized_rnn-inl.h (hunk 3 above)
Comment (bgawrych) [reply src/operator/quantization/quantized_rnn-inl.h:21]: it is used by two other .cc files, so yes

File: src/operator/quantization/dnnl/dnnl_quantized_rnn.cc (hunk 2 above)
Comment (bgawrych) [reply src/operator/quantization/dnnl/dnnl_quantized_rnn.cc]: done

**Review by anko-intel — APPROVED** [review]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [unix-cpu]

**Review by bartekkuncer — APPROVED** [review]
LGTM

**Comment by bgawrych** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [centos-cpu, unix-cpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [centos-cpu, unix-cpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [unix-cpu]
